### PR TITLE
feat(plugin): Support customplugin (#301)

### LINKED ISSUE #301: Support custom plugins for fluentbit
### Is your feature request related to a problem? Please describe.

fluentbit and fluentd have an unusually rich way of collecting and processing logs, mainly due to their multiple plugins. But fluent-operator does not integrate all plug-ins, and there should be a priority for the order of integrating plug-ins, which plug-ins do you want to integrate most?

### Describe the solution you'd like

We hope we will make suggestions and we will rank the plugin development integration according to priority.

### Describe alternatives you've considered

The integration of various plug-ins is integrated by customizing the CRD of the plug-ins, the steps required are roughly the same, and you are welcome to contribute code to improve the integration of plug-ins.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Gentleelephant**
Signed-off-by: Gentleelephant <birdhk@kubesphere.io>

### Which issue(s) this PR fixes:

Fixes #301 

### Does this PR introduced a user-facing change?

```release-note

```

### Additional documentation, usage docs, etc.:

```docs

```

### CODE REVIEW

File: apis/fluentbit/v1alpha2/plugins/custom/custom_types.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+			continue
+		}
+		if strings.EqualFold(tmp[0], "Name") {
+			return tmp[1]
+		}
+	}
+	return "NONE"
+}
+
+func (a *CustomPlugin) Params(_ plugins.SecretLoader) (*params.KVs, error) {
```
Comment (wanjunlei) [inline apis/fluentbit/v1alpha2/plugins/custom/custom_types.go:32]: Why need to parser the `config`? Why need to skip the `name` field?

**Review by wanjunlei — CHANGES_REQUESTED** [review]

File: apis/fluentbit/v1alpha2/plugins/custom/custom_types.go (hunk 1 above)
Comment (Gentleelephant) [reply apis/fluentbit/v1alpha2/plugins/custom/custom_types.go:33]: A custom plugin is also a plugin that implements the Plugin interface. The plugin name is already resolved in the Name method, so there is no need to resolve the Name again.

**Comment by benjaminhuo** [comment]
@Gentleelephant We also need to update docs for plugins https://github.com/fluent/fluent-operator/tree/master/docs/plugins/fluentbit 

@wanjunlei How to generate these docs automatically?

The index should be added manually as I can remember

**Comment by benjaminhuo** [comment]
> @Gentleelephant We also need to update docs for plugins https://github.com/fluent/fluent-operator/tree/master/docs/plugins/fluentbit
> 
> @wanjunlei How to generate these docs automatically?
> 
> The index should be added manually as I can remember

Make docs-update?

https://github.com/fluent/fluent-operator/blob/master/Makefile#L174

File: apis/fluentbit/v1alpha2/plugins/custom/custom_types.go (hunk 1 above)
Comment (wanjunlei) [reply apis/fluentbit/v1alpha2/plugins/custom/custom_types.go:33]: A more reasonable approach.

Add the `name` filed just when the `p.Name()` is not nil
https://github.com/fluent/fluent-operator/blob/master/apis/fluentbit/v1alpha2/clusterinput_types.go#L94



and make the `customPlugin.Name()` return nil

**Comment by wanjunlei** [comment]
> > @Gentleelephant We also need to update docs for plugins https://github.com/fluent/fluent-operator/tree/master/docs/plugins/fluentbit
> > @wanjunlei How to generate these docs automatically?
> > The index should be added manually as I can remember
> 
> make docs-update?
> 
> https://github.com/fluent/fluent-operator/blob/master/Makefile#L174

Yes

**Comment by Gentleelephant** [comment]
Regarding the modification of the document, do I need to create a new pr? @benjaminhuo @wanjunlei

**Comment by benjaminhuo** [comment]
@Gentleelephant You can add another doc in https://github.com/fluent/fluent-operator/tree/master/docs/best-practice to elaborate on how to use the custom plugin like docs in https://github.com/fluent/fluent-operator/blob/master/docs/best-practice/collect-systemd-logs.md

**Review by benjaminhuo — APPROVED** [review]

**Pull request merged**